Improve particle serialization scalability

gassmoeller opened the pull request:
Previously the particle serialization was not very scalable, because the data was serialized in the usual way for postprocessors. Every process had to send its data to process 0, which then wrote the data  into an output stream,  which was saved to disk. This is impractical or even impossible for large application models. The new implementation utilizes the same functionality that we use for remeshing, i.e. the particle data is attached to the cells of the triangulation and stored together with the mesh. In essence this is similar to a SolutionTransfer object for particle data. The drawback is that every cell has to store the same amount of data (p4est limitation), therefore we now need significantly more disk space. An example application with around 20 million particles now need 4 GB instead of 600 MB disk space for serialization. I still think the new approach is better, the old one simply did crash for large models.
Additionally the new approach was implemented by adding signals before and after serialization. This also allows other plugins to serialize data more efficiently. E.g. the code that handles the free surface serialization could now be moved outside of the checkpoint_restart file and into the free_surface file (I will open an issue for that).

bangerth commented:
Hm, hooking this into signals and letting each process handle it separately is clearly the way to go. The current serialization code can't scale.

At the same time, creating so much more output is a bit annoying. Writing 4GB of data is, even on parallel file systems, not fast. But let me look through the patch first :-)

bangerth commented on include/aspect/particle/world.h:
```diff
@@ -122,11 +122,14 @@ namespace aspect
+         * state variable, because it is needed to serialize and deserialize
+         * the particle data correctly in parallel (it determines the size of
+         * the data chunks per cell that are stored and read).
          */
```
can you add whether this state variable is always kept up to date? I guess the function is also public, which however no longer makes much sense since no external entity should call it I suppose.

bangerth commented on include/aspect/particle/world.h:
```diff
@@ -291,6 +298,13 @@ namespace aspect
         /**
+         * The maximum number of particles per cell in the global domain. This
+         * variable is important to store and load particle data during
+         * repartition and serialization of the solution.
```
same here -- is it always up to date?

bangerth commented on include/aspect/simulator_signals.h:
```diff
@@ -148,6 +146,33 @@ namespace aspect
+    * argument. This argument will point to the triangulation used by
+    * the Simulator class.
+    */
+    boost::signals2::signal<void (typename parallel::distributed::Triangulation<dim> &)>  pre_serialization_store_user_data;
```
since serialization can mean both storing and loading, maybe rename this and the next signal as `pre_store_...` and `post_load_...`?

bangerth commented on source/particle/world.cc:
```diff
@@ -163,26 +161,39 @@ namespace aspect
                                                                      std_cxx11::ref(*this),
+                                                                     false,
                                                                      std_cxx11::_1));
+      signals.pre_serialization_store_user_data.connect(std_cxx11::bind(&World<dim>::register_store_callback_function,
```
this is pre-existing, but I think the function you are attaching here is poorly named: "register_*" functions are ones that *connect a function X to a signal S*, but you actually name the function X itself this way. Maybe rename these functions `store_callback_function` and `load_callback_function`?

gassmoeller replied:
This is not quite correct, but I agree it is confusing. These functions are at the same time callback functions (to the SimulatorSignals), and `register...` functions, because their main purpose is to register the actual callback functions (called `store_tracers` and `load_tracers`) into the Triangulation by calling `register_data_attach` and `notify_ready_to_unpack`. I feel that `store_callback_function` is more misleading than their current name, but I am happy to rename them if we find a more descriptive name.

bangerth replied:
Ah, I see. OK then.

gassmoeller commented:
Not sure if you were finished with the review, but I addressed the comments that you had. Let me know if you are ok with the changes, then I can squash the commits (kept them for readability).

bangerth commented:
I think my only current issue is the problem with the restart file size. I don't have any other good solution, though. If you're convinced that this is currently the best solution, then we should go with it. (I think it would also have the advantage that we can restart with a different number of processors than we had when we checkpointed.)

gassmoeller commented:
Yes I do not see a good solution either. Still let us keep this PR open for a bit, I am having trouble of resuming large models even with the current improvements. Could be because of the file size of course, since all models with restart files smaller than 4 GB work fine, but I also found some above 4 GB to work. So I am a bit confused now, and debugging these large models is a pain.

bangerth commented:
OK. Feel free to merge once you are happy.
